### PR TITLE
Comments manager improvements/extensions needed by Talk

### PR DESCRIPTION
**Pull request by nickvergessen**
This is just upstreaming of what we already use in talk:
https://github.com/nextcloud/spreed/blob/ab443ee9f2dcc25e338d995b80faf5d65b8e4fac/lib/Chat/CommentsManager.php

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]
Sure! :)

File: lib/private/Comments/Manager.php [hunk 1]
```diff
@@ -440,11 +449,12 @@ public function getForObjectSince(
 								$query->createNamedParameter($lastKnownCommentDateTime, IQueryBuilder::PARAM_DATE),
 								IQueryBuilder::PARAM_DATE
 							),
-							$query->expr()->lt('id', $query->createNamedParameter($lastKnownCommentId))
+							$query->expr()->$idComparison('id', $query->createNamedParameter($lastKnownCommentId))
```
Comment (ChristophWurst) [inline lib/private/Comments/Manager.php:452]: wouldn't the ternary operator do a decent job as well? ``$includeLastKnown ? $query->expr()->lte(...) : $query->expr()->lt(...)``?

Small amount of duplication > string magic

File: lib/public/Comments/ICommentsManager.php [hunk 2]
```diff
@@ -164,6 +179,43 @@ public function search(string $search, string $objectType, string $objectId, str
...
+	 */
+	public function getLastCommentBeforeDate(string $objectType, string $objectId, \DateTime $beforeDate, string $verb = ''): int;
+
+	/**
+	 * @param string $objectType
+	 * @param string $objectId
+	 * @param string $verb
+	 * @param string $actorType
+	 * @param string[] $actors
+	 * @return array
```
Comment (ChristophWurst) [inline lib/public/Comments/ICommentsManager.php:208]: do we know the type?

File: lib/private/Comments/Manager.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/Comments/Manager.php:452]: I have mixed feelings in regards to multi-line ternary operators on multiline parameters for a function call. but we can do that outside. Let me change it

File: lib/public/Comments/ICommentsManager.php (hunk 2 above)
Comment (nickvergessen) [reply lib/public/Comments/ICommentsManager.php:208]: `string => \DateTime` :grin:

Comment (ChristophWurst) [reply lib/public/Comments/ICommentsManager.php:208]: `\DateTime[]` it is :P

File: lib/public/Comments/ICommentsManager.php [hunk 3]
```diff
@@ -205,7 +205,7 @@ public function getLastCommentBeforeDate(string $objectType, string $objectId, \
 	 * @param string $verb
 	 * @param string $actorType
 	 * @param string[] $actors
-	 * @return array
+	 * @return array Map of "string actor" => "\DateTime most recent comment date"
```
Comment (ChristophWurst) [inline lib/public/Comments/ICommentsManager.php:208]:
```suggestion
	 * @return \DateTime[] Map of "string actor" => "\DateTime most recent comment date"
```

Phpstorm and Psalm don't speak your language :)

Comment (ChristophWurst) [reply lib/public/Comments/ICommentsManager.php:208]: If you want to be fancy you can add an *additional* `@psalm-return array<string, \DateTime>`

**Comment by kesselb** [comment]
`lib/private/Comments/ManagerFactory.php:59:10:error - TooFewArguments: Too few arguments for OC\Comments\Manager::__construct - expecting 5 but saw 4`

To make Psalm happy ;)

**Comment by kesselb** [comment]
Drone is a bit unhappy about https://github.com/nextcloud/server/blob/master/tests/lib/Comments/FakeManager.php :fearful:

**Comment by kesselb** [comment]
Awesome @nickvergessen :tada: 

- Drone :heavy_check_mark: 
- Psalm :heavy_check_mark: 



The last thing to do: `@since or @deprecated tag is needed in PHPDoc for OCP\Comments\ICommentsManager::searchForObjects`

**Comment by nickvergessen** [comment]
Second review @kesselb @ChristophWurst

**Comment by nickvergessen** [comment]
Bump @kesselb @ChristophWurst @rullzer @blizzz

**Review by ChristophWurst — APPROVED** [review]

**Review by rullzer — APPROVED** [review]

**Pull request merged**